python/cudf/cudf/bindings/cudf_cpp.pyx

```diff
@@ -119,30 +215,39 @@ cdef gdf_scalar* gdf_scalar_from_scalar(val, dtype=None) except? NULL:
 
     set_scalar_value(s, val)
 
-    if dtype is not None:
-        s[0].dtype = dtypes[dtype.type]
-    else:
-        s[0].dtype = dtypes[val.dtype.type]
     s[0].is_valid = is_valid
+    s[0].dtype = gdf_dtype_from_value(val, dtype)
+
     return s
 
-cdef get_scalar_value(gdf_scalar scalar):
+
+cdef get_scalar_value(gdf_scalar scalar, dtype):
     """
     Returns typed value from a gdf_scalar
     0-dim array is retuned if dtype is date32/64, timestamp
     """
-    return {
-        GDF_FLOAT64: scalar.data.fp64,
-        GDF_FLOAT32: scalar.data.fp32,
-        GDF_INT64:   scalar.data.si64,
-        GDF_INT32:   scalar.data.si32,
-        GDF_INT16:   scalar.data.si16,
-        GDF_INT8:    scalar.data.si08,
-        GDF_BOOL8:   scalar.data.b08,
-        GDF_DATE32:  np.array(scalar.data.dt32).astype('datetime64[D]'),
-        GDF_DATE64:  np.array(scalar.data.dt64).astype('datetime64[ms]'),
-        GDF_TIMESTAMP: np.array(scalar.data.tmst).astype('datetime64[ns]'),
-    }[scalar.dtype]
+    if scalar.dtype == GDF_FLOAT64:
+        return scalar.data.fp64
+    if scalar.dtype == GDF_FLOAT32:
+        return scalar.data.fp32
+    if scalar.dtype == GDF_INT64:
+        return scalar.data.si64
+    if scalar.dtype == GDF_INT32:
+        return scalar.data.si32
+    if scalar.dtype == GDF_INT16:
+        return scalar.data.si16
+    if scalar.dtype == GDF_INT8:
+        return scalar.data.si08
+    if scalar.dtype == GDF_BOOL8:
+        return scalar.data.b08
+    if scalar.dtype == GDF_DATE32:
+        return np.array(scalar.data.dt32).astype(dtype)
+    if scalar.dtype == GDF_DATE64:
+        return np.array(scalar.data.dt64).astype(dtype)
+    if scalar.dtype == GDF_TIMESTAMP:
+        return np.array(scalar.data.tmst).astype(dtype)
```

Comment:
Instead of creating 0-dim numpy arrays here and typecasting should we just construct the proper numpy datetime scalar? I.E.
```python
return np.datetime64(scalar.data.tmst, dtype)
```

Reply:
Oh nice! Updated in https://github.com/rapidsai/cudf/pull/2213/commits/50f6e50cd507e70530631f536ddd3d1c2eca731a.